examples/online-exhibition/online-exhibition/03_ingest_table_assets.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import configparser
+import inspect
+import logging
+import os
+import requests
+
+import pandas as pd
+
+from vdk.api.job_input import IJobInput
+
+log = logging.getLogger(__name__)
+
+
+def run(job_input: IJobInput):
+    """
+    Download datasets required by the scenario and put them in the data lake.
+    """
+    log.info(f"Starting job step {__name__}")
+
+    
+    current_dir = os.path.dirname(os.path.abspath(__file__))
+    config_file = os.path.join(current_dir, 'config.ini')
+    print(config_file)
+
+    # url of the U.S. Life Expectancy at Birth by State and Census Tract - 2010-2015 dataset
+    #url = "http://data.cdc.gov/api/views/5h56-n989/rows.csv"
+    parser = configparser.ConfigParser()
+    parser.read(config_file)
+    api_key = parser.get("key", "api_key")
```

Comment:
Our current solution for keeping secretes is so called "Properties" https://github.com/vmware/versatile-data-kit/wiki/Features#keeping-state-configuration-and-secrets . 
config.ini is not really secret (it's part of the source code) so that's why it's not a good place.

You need to have [vdk-server](https://github.com/vmware/versatile-data-kit/wiki/Installation#install-locally) installed to work. then you can set them with `vdk properties --set-secret key` and access them in data job with `job_input.get_property()` method.

Reply:
I have set the key as follows:
```
vdk properties --set-secret api_key <RETURN>

Job Name: online-exhibition
Job Team: my-team
api_key: MY_API_KEY
```
I have updated the code but it still does not work, because it gives me the following error:

```
angelica@MacBook-Air-di-Angelica online-exhibition % vdk run online-exhibition
Versatile Data Kit (VDK)
Version: 0.0.461022384
Build details: RELEASE_VERSION=0.0.461022384, BUILD_DATE=Tue Feb  1 16:01:47 UTC 2022, BUILD_MACHINE_INFO=Linux runner-opyxurxq-project-28359933-concurrent-02shxt 5.4.129-62.227.amzn2.x86_64 #1 SMP Wed Jul 7 00:08:43 UTC 2021 x86_64 GNU/Linux, GITLAB_CI_JOB_ID=2040549026, GIT_COMMIT_SHA=86dd875717e4b711df7b6f25a5ea08e6f0de438e, GIT_BRANCH=main
Python version: 3.8.10 64bit (/Library/Frameworks/Python.framework/Versions/3.8/bin/python3.8)

Installed plugins:
vdk-ingest-file (from package vdk-ingest-file, version 0.1.415625538)
vdk-ingest-http (from package vdk-ingest-http, version 0.1.428971094)
vdk-control-service-properties (from package vdk-plugin-control-cli, version 0.1.417315215)
vdk-execution-skip (from package vdk-plugin-control-cli, version 0.1.417315215)
vdk-plugin-control-cli (from package vdk-plugin-control-cli, version 0.1.417315215)
vdk-csv (from package vdk-csv, version 0.1.377908503)
vdk-trino (from package vdk-trino, version 0.1.460149153)
vdk-server (from package vdk-server, version 0.1.424970629)
vdk-sqlite (from package vdk-sqlite, version 0.1.415630020)
--------------------------------------------------------------------------------
Run job with directory /Users/angelica/CNR/Git/versatile-data-kit/examples/online-exhibition/online-exhibition
2022-02-25 11:46:17,111 [VDK] online-exhibition [INFO ] vdk.plugin.control_cli_plugin. properties_plugin.py:30   initialize_job  [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Control Service REST API URL is not configured. Will not initialize Control Service based Properties client implementation.
2022-02-25 11:46:17,111 [VDK] online-exhibition [INFO ] vdk.plugin.control_cli_plugin.    execution_skip.py:105  _skip_job_if_nec[id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Checking if job should be skipped:
2022-02-25 11:46:17,111 [VDK] online-exhibition [INFO ] vdk.plugin.control_cli_plugin.    execution_skip.py:106  _skip_job_if_nec[id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Job : online-exhibition, Team : my-team, Log config: LOCAL, execution_id: 365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977
2022-02-25 11:46:17,111 [VDK] online-exhibition [INFO ] root                              execution_skip.py:111  _skip_job_if_nec[id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Local execution, skipping parallel execution check.
2022-02-25 11:46:17,112 [VDK] online-exhibition [INFO ] vdk.internal.builtin_plugins.r         job_input.py:87   _substitute_quer[id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Data Job Properties has not been initialized., so I won't be able to provide query properties substitution capabilities from job properties.If passed job arguments will still be used
2022-02-25 11:46:17,112 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:123  execute         [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Executing query:
-- job_name: online-exhibition
-- op_id: 365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977
DROP TABLE IF EXISTS assets

2022-02-25 11:46:17,121 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:126  execute         [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Executing query SUCCEEDED.
2022-02-25 11:46:17,121 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:147  fetchall        [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Fetching all results from query ...
2022-02-25 11:46:17,133 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:150  fetchall        [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Fetching all results from query SUCCEEDED.
2022-02-25 11:46:17,133 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:157  close           [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Closing DB cursor ...
2022-02-25 11:46:17,133 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:159  close           [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Closing DB cursor SUCCEEDED.
2022-02-25 11:46:17,134 [VDK] online-exhibition [INFO ] vdk.internal.builtin_plugins.r         job_input.py:87   _substitute_quer[id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Data Job Properties has not been initialized., so I won't be able to provide query properties substitution capabilities from job properties.If passed job arguments will still be used
2022-02-25 11:46:17,134 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:123  execute         [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Executing query:
-- job_name: online-exhibition
-- op_id: 365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977
 select 1 -- Testing if connection is alive.
2022-02-25 11:46:17,136 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:126  execute         [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Executing query SUCCEEDED.
2022-02-25 11:46:17,136 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:123  execute         [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Executing query:
-- job_name: online-exhibition
-- op_id: 365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977
CREATE TABLE assets (
    completeness int,
    country varchar,
    dataProvider varchar,
    dcCreator varchar,
    dcCreatorLangAware varchar,
    dcTitleLangAware varchar,
    edmDatasetName varchar,
    edmIsShownBy varchar,
    edmPreview varchar,
    edmTimespanLabel varchar,
    edmTimespanLabelLangAware varchar,
    europeanaCollectionName varchar,
    europeanaCompleteness int,
    "guid" varchar,
    id varchar,
    index int,
    language varchar,
    link varchar,
    previewNoDistribute boolean,
    provider varchar,
    rights varchar,
    score double,
    "timestamp" varchar,
    timestamp_created varchar,
    timestamp_created_epoch varchar,
    timestamp_update varchar,
    timestamp_update_epoch varchar,
    title varchar,
    "type" varchar,
    ugc varchar
)

2022-02-25 11:46:17,138 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:126  execute         [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Executing query SUCCEEDED.
2022-02-25 11:46:17,138 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:147  fetchall        [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Fetching all results from query ...
2022-02-25 11:46:17,178 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:150  fetchall        [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Fetching all results from query SUCCEEDED.
2022-02-25 11:46:17,178 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:157  close           [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Closing DB cursor ...
2022-02-25 11:46:17,178 [VDK] online-exhibition [INFO ] vdk.plugin.trino.trino_connect    managed_cursor.py:159  close           [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Closing DB cursor SUCCEEDED.
2022-02-25 11:46:17,476 [VDK] online-exhibition [INFO ] vdk.internal.builtin_plugins.r   file_based_step.py:82   run_python_step [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Entering 03_ingest_table_assets.py#run(...) ...
2022-02-25 11:46:17,476 [VDK] online-exhibition [INFO ] step_03_ingest_table_assets    03_ingest_table_asse:19   run             [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Starting job step step_03_ingest_table_assets
/Users/angelica/CNR/Git/versatile-data-kit/examples/online-exhibition/online-exhibition/config.ini
2022-02-25 11:46:17,479 [VDK] online-exhibition [ERROR] vdk.internal.builtin_plugins.j            errors.py:291  log_and_throw   [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- A configuration error  occurred. The error should be resolved by Platform error. Here are the details:
  WHAT HAPPENED : I'm trying to call method 'get_property' and failed.
WHY IT HAPPENED : No properties client implementation has been installed or configured.
   CONSEQUENCES : Current  Step will fail, and as a result the whole Data Job will fail.
COUNTERMEASURES : Check if you have installed properly vdk. Check if you need to install some plugin for properties. If a plugin is installed, check out plugin documentation whether the plugin is configured correctly.
Traceback (most recent call first):
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/job_properties/properties_router.py", line 123, in <lambda>
    error_handler = lambda methodname: errors.log_and_throw(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/job_properties/Propertiesnotavailable.py", line 27, in tell_user
    self._error_handler(methodname)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/job_properties/Propertiesnotavailable.py", line 18, in get_property
    self.tell_user("get_property")
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/run/job_input.py", line 69, in get_property
    return self.__properties_router.get_properties_impl().get_property(
  File "/Users/angelica/CNR/Git/versatile-data-kit/examples/online-exhibition/online-exhibition/03_ingest_table_assets.py", line 30, in run
    api_key = job_input.get_property("key")
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/run/file_based_step.py", line 117, in invoke_run_function
    func(**actual_arguments)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/run/file_based_step.py", line 83, in run_python_step
    StepFuncFactory.invoke_run_function(func, job_input)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/run/data_job.py", line 71, in run_step
    step_executed = step.runner_func(step, context.job_input)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/run/data_job.py", line 131, in run_job
    res = context.core_context.plugin_registry.hook().run_step(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/run/data_job.py", line 289, in run
    return self._plugin_hook.run_job(context=job_context)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/run/cli_run.py", line 64, in create_and_run_data_job
    execution_result = job.run(args)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/builtin_plugins/run/cli_run.py", line 117, in run
    run_impl.create_and_run_data_job(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/cli_entry.py", line 95, in vdk_cli_execute
    exit_code = root_command(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/cli_entry.py", line 132, in vdk_main
    exit_code = cast(InternalHookSpecs, plugin_registry.hook()).vdk_cli_execute(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/manager.py", line 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/pluggy/hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vdk/internal/cli_entry.py", line 167, in main
    exit_code = cast(InternalHookSpecs, plugin_registry.hook()).vdk_main(
  File "/Library/Frameworks/Python.framework/Versions/3.8/bin/vdk", line 8, in <module>
    sys.exit(main())

2022-02-25 11:46:17,479 [VDK] online-exhibition [ERROR] vdk.internal.builtin_plugins.r   file_based_step.py:90   run_python_step [id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Exiting  03_ingest_table_assets.py#run(...) FAILURE
2022-02-25 11:46:17,479 [VDK] online-exhibition [INFO ] vdk.internal.builtin_plugins.r           cli_run.py:65   create_and_run_d[id:365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977-ecd37]- Data Job execution summary: {
  "data_job_name": "online-exhibition",
  "execution_id": "365e0dfb-0b6c-43f6-b343-c9e3cbf4cdc4-1645785977",
  "start_time": "2022-02-25T10:46:17.111842",
  "end_time": "2022-02-25T10:46:17.479739",
  "status": "error",
  "steps_list": [
    {
      "name": "01_delete_table_assets.sql",
      "type": "sql",
      "start_time": "2022-02-25T10:46:17.111863",
      "end_time": "2022-02-25T10:46:17.133598",
      "status": "success",
      "details": null,
      "exception": null,
      "blamee": null
    },
    {
      "name": "03_create_table_assets.sql",
      "type": "sql",
      "start_time": "2022-02-25T10:46:17.133681",
      "end_time": "2022-02-25T10:46:17.178668",
      "status": "success",
      "details": null,
      "exception": null,
      "blamee": null
    },
    {
      "name": "03_ingest_table_assets.py",
      "type": "python",
      "start_time": "2022-02-25T10:46:17.178734",
      "end_time": "2022-02-25T10:46:17.479696",
      "status": "error",
      "details": "An exception occurred, exception message was: A configuration error  occurred. The error should be resolved by Platform error. Here are the details:\n  WHAT HAPPENED : I'm trying to call method 'get_property' and failed.\nWHY IT HAPPENED : No properties client implementation has been installed or configured.\n   CONSEQUENCES : Current  Step will fail, and as a result the whole Data Job will fail.\nCOUNTERMEASURES : Check if you have installed properly vdk. Check if you need to install some plugin for properties. If a plugin is installed, check out plugin documentation whether the plugin is configured correctly.",
      "exception": {
        "message": "A configuration error  occurred. The error should be resolved by Platform error. Here are the details:\n  WHAT HAPPENED : I'm trying to call method 'get_property' and failed.\nWHY IT HAPPENED : No properties client implementation has been installed or configured.\n   CONSEQUENCES : Current  Step will fail, and as a result the whole Data Job will fail.\nCOUNTERMEASURES : Check if you have installed properly vdk. Check if you need to install some plugin for properties. If a plugin is installed, check out plugin documentation whether the plugin is configured correctly.",
        "is_logged": true
      },
      "blamee": "Configuration error",
      "exception_name": "VdkConfigurationError"
    }
  ],
  "exception": null
}
Error: A configuration error  occurred. The error should be resolved by Platform error. Here are the details:
  WHAT HAPPENED : I'm trying to call method 'get_property' and failed.
WHY IT HAPPENED : No properties client implementation has been installed or configured.
   CONSEQUENCES : Current  Step will fail, and as a result the whole Data Job will fail.
COUNTERMEASURES : Check if you have installed properly vdk. Check if you need to install some plugin for properties. If a plugin is installed, check out plugin documentation whether the plugin is configured correctly.
```